File: src/main/java/com/sraddons/mixin/CommandSuggestionsMixin.java
package com.sraddons.mixin;

import com.mojang.brigadier.ParseResults;
import com.mojang.brigadier.StringReader;
import com.mojang.brigadier.context.StringRange;
import com.mojang.brigadier.suggestion.Suggestions;
import com.mojang.brigadier.suggestion.Suggestion;
import com.sraddons.config.SRConfig;
import com.sraddons.feature.partycommands.commands.Commands;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.components.CommandSuggestions;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.commands.SharedSuggestionProvider;
import org.spongepowered.asm.mixin.Final;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

@Mixin(CommandSuggestions.class)
public abstract class CommandSuggestionsMixin {
    @Shadow
    private ParseResults<SharedSuggestionProvider> currentParse;

    @Shadow
    @Final
    private EditBox input;

    @Shadow
    private boolean keepSuggestions;

    @Shadow
    private CompletableFuture<Suggestions> pendingSuggestions;

    @Shadow
    private CommandSuggestions.SuggestionsList suggestions;

    @Shadow
    private boolean allowSuggestions;

    @Shadow
    private List<net.minecraft.util.FormattedCharSequence> commandUsage;

    @Shadow
    protected abstract void showSuggestions(boolean bl);

    @Shadow
    private void updateUsageInfo() {}

    @Inject(method = "showSuggestions", at = @At("HEAD"))
    private void onShowSuggestions(boolean bl, CallbackInfo ci) {
        String value = this.input.getValue();
        String prefix = SRConfig.INSTANCE.getSettings().getPartyCommands().getPrefix();

        if (value.startsWith(prefix) && this.pendingSuggestions != null && this.pendingSuggestions.isDone()) {
            try {
                Suggestions suggestions = this.pendingSuggestions.getNow(null);
                if (suggestions != null) {
                    List<Suggestion> filtered = suggestions.getList().stream()
                        .filter(s -> !s.getText().startsWith("!"))
                        .collect(Collectors.toList());

                    if (filtered.size() != suggestions.getList().size()) {
                        this.pendingSuggestions = CompletableFuture.completedFuture(
                            new Suggestions(suggestions.getRange(), filtered)
                        );
                    }
                }
            } catch (Exception ignored) {}
        }
    }

    @Inject(method = "updateCommandInfo", at = @At("HEAD"), cancellable = true)
    private void onUpdateCommandInfo(CallbackInfo ci) {
        String value = this.input.getValue();
        String prefix = SRConfig.INSTANCE.getSettings().getPartyCommands().getPrefix();
        int length = prefix.length();

        if (value.startsWith(prefix)) {
            if (!this.keepSuggestions) {
                this.input.setSuggestion(null);
                this.suggestions = null;
            }

            if (this.currentParse != null && !this.currentParse.getReader().getString().equals(value)) {
                this.currentParse = null;
            }

            StringReader reader = new StringReader(value);
            reader.setCursor(length);

            if (this.currentParse == null) {
                var player = Minecraft.getInstance().player;
                if (player != null) {
                    this.currentParse = Commands.DISPATCHER.parse(reader, player.connection.getSuggestionsProvider());
                }
            }

            int cursor = this.input.getCursorPosition();
            if (cursor >= length && (this.suggestions == null || !this.keepSuggestions)) {
                if (this.currentParse != null && !this.currentParse.getExceptions().isEmpty()) {
                    this.pendingSuggestions = CompletableFuture.completedFuture(
                        new Suggestions(StringRange.at(cursor), Collections.emptyList())
                    );
                } else {
                    this.pendingSuggestions = Commands.DISPATCHER.getCompletionSuggestions(this.currentParse, cursor);
                }

                this.pendingSuggestions = this.pendingSuggestions.thenApply(suggestions -> {
                    List<Suggestion> filtered = suggestions.getList().stream()
                        .filter(s -> !s.getText().startsWith("!"))
                        .collect(Collectors.toList());
                    return new Suggestions(suggestions.getRange(), filtered);
                });

                this.pendingSuggestions.thenRun(() -> {
                    if (this.pendingSuggestions.isDone() && this.allowSuggestions) {
                        if (this.currentParse != null && this.currentParse.getReader().getString().equals(value)) {
                            this.showSuggestions(false);
                        }
                    }
                });
            }

            if (this.currentParse != null && !this.currentParse.getExceptions().isEmpty()) {
                this.updateUsageInfo();
            } else {
                this.commandUsage.clear();
            }

            ci.cancel();
        }
    }
}

File: src/main/kotlin/com/sraddons/mixin/CommandSuggestionsMixin.kt
package com.sraddons.mixin

import com.mojang.brigadier.ParseResults
import com.mojang.brigadier.StringReader
import com.mojang.brigadier.context.StringRange
import com.mojang.brigadier.suggestion.Suggestions
import com.sraddons.config.SRConfig
import com.sraddons.feature.partycommands.commands.Commands
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.components.CommandSuggestions
import net.minecraft.client.gui.components.EditBox
import net.minecraft.client.multiplayer.ClientSuggestionProvider
import net.minecraft.util.FormattedCharSequence
import org.spongepowered.asm.mixin.Final
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import java.util.concurrent.CompletableFuture

@Mixin(CommandSuggestions::class)
abstract class CommandSuggestionsMixin {

    @Shadow
    @JvmField
    var currentParse: ParseResults<ClientSuggestionProvider>? = null

    @Shadow
    @Final
    @JvmField
    var input: EditBox? = null

    @Shadow
    @JvmField
    var keepSuggestions: Boolean = false

    @Shadow
    @JvmField
    var pendingSuggestions: CompletableFuture<Suggestions>? = null

    @Shadow
    @JvmField
    var suggestions: CommandSuggestions.SuggestionsList? = null

    @Shadow
    @JvmField
    var allowSuggestions: Boolean = false

    @Shadow
    @JvmField
    var commandUsage: List<FormattedCharSequence>? = null

    @Shadow
    protected abstract fun showSuggestions(bl: Boolean)

    @Shadow
    private fun updateUsageInfo(parse: ParseResults<ClientSuggestionProvider>, suggestions: Suggestions?) {}

    @Inject(method = ["showSuggestions"], at = [At("HEAD")])
    private fun onShowSuggestions(bl: Boolean, ci: CallbackInfo) {
        val editBox = input ?: return
        val pending = pendingSuggestions ?: return
        val value = editBox.value
        val prefix = SRConfig.settings.partyCommands.prefix

        if (value.startsWith(prefix) && pending.isDone) {
            try {
                val brigadierSuggestions = pending.getNow(null)
                if (brigadierSuggestions != null) {
                    val filtered = brigadierSuggestions.list.filter { !it.text.startsWith("!") }

                    if (filtered.size != brigadierSuggestions.list.size) {
                        pendingSuggestions = CompletableFuture.completedFuture(
                            Suggestions(brigadierSuggestions.range, filtered)
                        )
                    }
                }
            } catch (e: Exception) {
                org.apache.logging.log4j.LogManager.getLogger("SR-Addons-Mixin")
                    .warn("Failed to filter ! suggestions", e)
            }
        }
    }

    @Inject(method = ["updateCommandInfo"], at = [At("HEAD")], cancellable = true, require = 1)
    private fun onUpdateCommandInfo(ci: CallbackInfo) {
        val editBox = input ?: return
        val value = editBox.value
        val prefix = SRConfig.settings.partyCommands.prefix
        val length = prefix.length

        if (!value.startsWith(prefix)) return

        if (!keepSuggestions) {
            editBox.setSuggestion(null)
            suggestions = null
        }

        if (currentParse?.reader?.string != value) {
            currentParse = null
        }

        val reader = StringReader(value)
        reader.cursor = length

        if (currentParse == null) {
            val player = Minecraft.getInstance().player
            if (player != null) {
                val raw = Commands.DISPATCHER.parse(reader, player.connection.suggestionsProvider)
                currentParse = @Suppress("UNCHECKED_CAST") (raw as? ParseResults<ClientSuggestionProvider>) ?: return
            }
        }

        val cursor = editBox.cursorPosition
        if (cursor >= length && (suggestions == null || !keepSuggestions)) {
            val parse = currentParse
            if (parse != null && parse.exceptions.isNotEmpty()) {
                pendingSuggestions = CompletableFuture.completedFuture(
                    Suggestions(StringRange.at(cursor), emptyList())
                )
            } else {
                @Suppress("UNCHECKED_CAST")
                pendingSuggestions = Commands.DISPATCHER.getCompletionSuggestions(
                    parse as? ParseResults<net.minecraft.commands.SharedSuggestionProvider>, cursor)
            }

            pendingSuggestions = pendingSuggestions?.thenApply { s ->
                val filtered = s.list.filter { !it.text.startsWith("!") }
                Suggestions(s.range, filtered)
            }

            if (pendingSuggestions?.isDone == true && allowSuggestions) {
                if (currentParse?.reader?.string == value) {
                    showSuggestions(false)
                }
            }
        }

        val parse = currentParse
        if (parse?.exceptions?.isNotEmpty() == true) {
            updateUsageInfo(parse, pendingSuggestions?.getNow(null))
        } else {
            (commandUsage as? MutableList<*>)?.clear()
        }

        ci.cancel()
    }
}
